types/node/globals.d.ts

```diff
@@ -53,10 +53,26 @@ declare global {
     // Same as module.exports
     var exports: any;
 
+    interface GCFunction {
+        (options: {
+            execution?: "sync";
+            flavor?: "regular" | "last-resort";
+            type?: "major-snapshot" | "major" | "minor";
+            filename?: string;
+        }): void;
+        (options: {
+            execution: "async";
+            flavor?: "regular" | "last-resort";
+            type?: "major-snapshot" | "major" | "minor";
+            filename?: string;
+        }): Promise<void>;
```

Comment:
```suggestion
        (options: {
            execution?: "sync" | undefined;
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): void;
        (options: {
            execution: "async";
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): Promise<void>;
```
Apologies for not including previously - this is required for compatibility with `exactOptionalPropertyTypes`.

Reply:
> Apologies for not including previously - this is required for compatibility with `exactOptionalPropertyTypes`.

How so? Why should `undefined` be accepted?

My opinion is that it would be wrong - if the user actually has `exactOptionalPropertyTypes` set, they regard `undefined`-valued properties as a mistake, and we should respect that.

For background, the value `undefined` is [treated internally](https://github.com/nodejs/node/blob/d65b17082bd6b559fc6666e14ef5bd0079389ecc/deps/v8/src/extensions/gc-extension.cc#L54) the same as passing a property that is not a string. That is, for all arguments, it's tolerated but ignored, just as if passing `null` or even `42`.

Reply:
Indeed, and it's the convention in this repo to type options interfaces in that way ([example](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b4ae44884a5a4a56d6059d81f1567d3f6983dbe8/types/node/zlib.d.ts#L96-L123)) to allow for "positively" specifying the default parameter.

Reply:
It seems like that was a change to preserve backwards compatibility, but since this declaration did not precede TS 4.4, there is no backwards compatibility to maintain.

It makes the code less correct and less readable. If you want to make that change in another PR, I can't stop you.

https://github.com/microsoft/TypeScript/issues/44419#issuecomment-874662681

> The key with these automated PRs is that they are 'backwards' compatible in the sense that 4.4 behavior with exactOptionalPropertyTypes enabled is gives you the same type system behavior as 4.3 without it.  ... However, post the PRs it now easy to switch the semantics to be more accurate on a case by case basis.

Reply:
I believe the meta is still to encode `undefined` as the semantic value for "default" (allows for overriding to default in merged objects etc.) but I don't want to block this from being reviewed.

Reply:
Oops, didn't see this thread.

You definitely strongly recommend putting `| undefined` on these. You have the backwards compat flipped; this is for 4.4 and _newer_. If you use this API today and pass in an undefined param (maybe you want to conditionally do something), it won't work.

Reply:
No, I think I understand the compatibility reason.

AFAICT, the compatibility reason is so that code which worked pre-4.4 with the old behavior would still pass type-checking post-4.4.

It’s not needed since the declaration did not exist before Typescript introduced the breaking change to the `?` modifier.

For merged objects, `undefined` behaves very differently from function calls, so it’s better to *not* confuse the two at the type level all else being equal.

This will print 1 `((x=1)=>{console.log(x)})(undefined)`

This will print `undefined`: `console.log(Object.assign({x:1}, {x:undefined}).x)`

Reply:
Without `| undefined`, this code is not legal:

```
global.gc({ execution: isSync : "sync" : undefined })
```

Reply:
> Without `| undefined`, this code is not legal:
> 
> ```
> global.gc({ execution: isSync : "sync" : undefined })
> ```

Nor should it be!

1. Union types don't distribute over overloaded function calls so this won't work anyway (assuming you meant `global.gc({ execution: isAsync ? "async" : undefined })`)

2. If you pass `global.gc({})` or `global.gc({execution:undefined})`, it will default to the legacy "truthy parameter" behavior of a *minor*, synchronous GC pass. So you *don't* want conditionally-undefined properties for this API.

3. `| undefined` is wordier. Better to keep things short and sweet.

Reply:
I chose a bad property, clearly; let me try again: [Playground Link](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/PTAEGUHcEMAdQKYA9oGMAuB5W6CWB7AO2gBsAFAJ31gQvQE8AVemgZ1ACMF1IEFDQ6CgFcEoaIQAmoAGalWCAFwAoZSFAABZGiw4CxclRp0mLBK0WCRCVbkLpac1GIDiAYQBiwwhn2gA3sqgwaAAFNR4RBYBQSFxyAiowpGEAPyWAESs9D4ZANyxccEyJNAAbvgU6aAZFAgA5sKlFBmgAD41pazoALR1rJXo+YVFDDTVGQC20ABWlT2sxLCsABb4Q+0103Mtm1N2lcNFITK4JPzQkwjV3RR29QVxAL4AlJYVuJKPIeF6UZaBY7BBJJFKZaDZXLfY4lcqVCZ1RrNVodDJdXr9QZHIFja6ZbbzRZwVbrFFbWaHPaTA4taFFU7nYhXG5Ce5016WSj4akKAA8H0kAD4Ck9VJJEs0xGVoBRQPVUJZ3F4fCkCsoZN5fERQJJ8C5UKFcKwAEoNJoyywcfD4c4SAA0sjOFyullu9xeMTi8tCgKBsIqFEsRtNSJloFSNUR5t2lm84tOhAQkjtIziDOdCBTzxeyieQA)



> 3\. `| undefined` is wordier. Better to keep things short and sweet.

Wordiness or not, adding it changes the meaning.

Reply:
> I chose a bad property, clearly; let me try again

Since "regular" is the default, I expect you mean to toggle between flavor="regular" vs "last-resort". And that filename is not intended since this only has an effect for type="major-with-snapshot".

So I'll respond as if you have written this:
```ts
function doGc(isLastResort: boolean) {
    gc({
        flavor: isLastResort ? "last-resort" : undefined,
    })
}
```

Because of how this sets `found_options_object` in [gc-extension.cc](https://github.com/nodejs/node/blob/3a5b5428c6509f91306e10486f1e56e3b1e80195/deps/v8/src/extensions/gc-extension.cc), calling `doGc(true)` would result in a major, last-resort gc pass. But `doGc(false)` would result in a MINOR gc pass, which is not what's intended.

You should instead write:
```ts
function doGc(isLastResort: boolean) {
    gc({
        flavor: isLastResort ? "last-resort" : "regular",
    })
}
```
so that `doGc(false)` will result in a MAJOR gc pass, as probably intended.

---

> Wordiness or not, adding it changes the meaning.

100% agreed. Which is why it shouldn't be added - it makes the user has to ask "is there a difference between these two possible inputs"? It's also why I didn't add `| null | "" | 0` to every property, even though these result in the exact same behavior that `undefined` would.

If the user wants to be strict about "exactOptionalPropertyTypes" (which IMO is very reasonable), then blindly adding `| undefined` to all APIs means that TypeScript *won't* catch accidentally passing `undefined` to the property. It defeats the of having such a check in the first place!

Reply:
I am really not in a position to make claims about what particular inputs do or whether or not my specific code "makes sense"; I am more concerned about the types and ensuring DT consumers get working types. I can just as easily produce an example that doesn't use `gc` to avoid getting in the weeds about this particular API.

> If the user wants to be strict about "exactOptionalPropertyTypes" (which IMO is very reasonable), then blindly adding `| undefined` to all APIs means that TypeScript _won't_ catch accidentally passing `undefined` to the property. It defeats the of having such a check in the first place!

The difference is just about whether `gc({})` and `gc({ flavor: undefined })` behave the same way. If they do, then `| undefined` should be there. I'm not sure how I can explain this more clearly.

There are some APIs where this does matter. It's not clear that it does for `gc`.

Reply:
> I am more concerned about the types and ensuring DT consumers get working types.

What do you mean by "working types"? This type works as intended. The reason to add `| undefined` was so that transitioning to 4.4 wouldn't introduce new breakages for consumers. Not a concern for this PR, since TS 4.4 is already out there.

> The difference is just about whether `gc({})` and `gc({ flavor: undefined })` behave the same way. If they do, then `| undefined` should be there. I'm not sure how I can explain this more clearly.

`gc({})`, `gc(1)`, `gc({ flavor: undefined })`, `gc({ flavor: null })`, `gc({ flavor: "vanilla" })` *all* behave the same way; if I could prohibit them all at the type level (in an idiomatic and readable way), I would.

> There are some APIs where this does matter. It's not clear that it does for `gc`.

Agreed. That's why I omitted it - same as the reason I omitted other values that add no expressive power to the API. Even `null` which has the analogous "strictNullChecks" option.

----

A better pattern for defining a conditional property on an object literal which DOESN'T require the consuming function to weaken its type signature:

```ts
someFunction({
    key: value,
    ... (condition1 ? {key1: value1} : {key2: value2}),
    ... (condition2 && {key3: value3}),
})
```
----

Why do you think explicitly allowing `undefined` as a property value is a *good* idea?

Reply:
> The reason to add `| undefined` was so that transitioning to 4.4 wouldn't introduce new breakages for consumers. Not a concern for this PR, since TS 4.4 is already out there.

This continues to be backwards. This is all about users of the flag.



> Agreed. That's why I omitted it - same as the reason I omitted other values that add no expressive power to the API.

You've flipped the meaning of my statement; the fact that it doesn't matter is a reason _to_ include `| undefined`.

> Even `null` which has the analogous "strictNullChecks" option.

That's not what that option means... That flag controls the existence of `undefined` and `null` both.



> Why do you think explicitly allowing `undefined` as a property value is a _good_ idea?

Because that pattern is not on its face better; needing to control the exact shape of an options bag when the API simply does not care about it is precisely _why_ `| undefined` is added to people's types.

Types are never "always right" or "always wrong", it's a scale of usibility. Plenty of code out there could accept garbage and "behave" in a defined way, but the types make the choice to disallow constructs that are probably undesirable, passing in `1`, `"vanilla"`, etc, as you've described. I simply do not think that disallowing the passing of `| undefined` is the right place to draw the line.

You seem impossible to convince of this, though, so I'll just drop it. This package is the most critical DT package, hence trying to ensure that the right changes are made.

Reply:
> You've flipped the meaning of my statement; the fact that it doesn't matter is a reason _to_ include `| undefined`.

By adding `| undefined` to signatures, you're encouraging people to be sloppy when it adds no expressiveness.

>> Even `null` which has the analogous "strictNullChecks" option.

> That's not what that option means... That flag controls the existence of undefined and null both.

The same justification for adding `| undefined` suggests adding `| null`.
strictNullChecks allows sloppy code in the same way

```ts
gc({
  flavor: isLastResort ? "last-resort" : null,
})
```

[Playground link](https://www.typescriptlang.org/play/?strictNullChecks=false&exactOptionalPropertyTypes=true#code/PTAEGUHcEMAdQM4BcBOBLAxkgcgVwDb4DCAFgKYYDWCoARmUpGWQHaiq5mjQsAmoAM2j4EZAFwAoCSFAABZOix5CpCtTGDhoqWhZIyKIRi4BxIgDFcLLGgD2bAN4TQL0AApbsJHZYINT10DQMgAPClxvewB+DQAiBABPa1iAbmcglwF8aAA3WxQY0FiUMgBzAmgUWNAAHyLs5ABaEoR8pFT0jKQE2DJC2IBbaAArfMaEFjgEElt22qKh0ar5wd18jozMtHxWaAG+jQVdUrTAgF8ASg08tF5T1w8vHz9QAM3Q8MiWOOhE5PuMllcvl+iVytllnVYg0kM0yK0UO0AUFur1+osxhMpjM5lCMZCFmsqsjAgJtrt9oUjiwTp1QJcNAAFFC2AZoUQAHhuvAAfGkzlJeBQIVwcpVQKUMBozJZrF80hIBFYbPZQLxbCYMG52QAZX5IABK8LaGlotlsOx4ABpBOTJvtDqhjhdXnTJW43oDsnkUBpdfqjQikKAovV9XCg9UNCwCPgrXTLhIzkA)

>  I simply do not think that disallowing the passing of `| undefined` is the right place to draw the line.

I'm not "disallowing" passing `| undefined`. I'm just not adding it since there's little benefit in readability or usability.